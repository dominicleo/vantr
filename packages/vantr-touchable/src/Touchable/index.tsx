import * as React from 'react'
import View from '../View'
import convertRef from '../View/convertRef'
import { TouchablePropType } from '../PropsType'
import usePressEvents from '../usePressEvents'

const delayPressIn = 50
const delayPressOut = 0
const delayLongPress = 450

const Touchable: React.FC<TouchablePropType> = props => {
  const {
    onPress,
    onLongPress,
    onPressIn,
    onPressOut,

    disabled,
    children,
  } = props

  // 将没有交互反馈的元素都视为 disabled ，这样子才不会对上层 Touchable 元素产生影响
  const sysDisabled =
    disabled || (!onLongPress && !onPress && !onPressIn && !onPressOut)

  const pressConfig = React.useMemo(
    () => ({
      delayLongPress,
      delayPressIn,
      delayPressOut,
      disabled: sysDisabled,
      onLongPress,
      onPress: event => {
        // 类型为 touchend 表示是 touch 集合的 onPress
        if (event.type === 'touchend') {
          // 由于 onPress(touchend) 比 click input focus 等事件早执行，为了避免一些穿透现象，这里阻止了 onPress 的默认行为
          event.preventDefault()
          // 由于上一行取消了 onPress (touchend) 的默认行为，即按钮不会再触发 click 事件，这里做一个 click 事件补偿，以免上层元素有 click 相关的事件（比如自动埋点）没办法触发
          event.target?.click?.()
        }
        onPress?.(event)

        // TODO: 类型为 mouseup 表示为 mouse 集合的 onPress
        // 目前这一类并没有点击穿透的现象，因此这里不做特殊处理
      },
      onPressIn,
      onPressOut,
    }),
    [sysDisabled, onLongPress, onPress, onPressIn, onPressOut],
  )

  const child = React.Children.only(children) as any
  const { hostRef, ref } = convertRef(child?.ref)
  const pressEventHandlers = usePressEvents(hostRef, pressConfig)

  return (
    <View {...pressEventHandlers}>
      {React.cloneElement(child, {
        ref,
      })}
    </View>
  )
}

export default Touchable
