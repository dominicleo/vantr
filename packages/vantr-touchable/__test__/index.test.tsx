import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Touchable from '../src';

describe('Touchable', () => {
  it('snapshot', () => {
    const component = render(
      <Touchable>
        <div />
      </Touchable>,
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('onPress', () => {
    const fn = jest.fn();
    const component = render(
      <Touchable onPress={fn}>
        <div />
      </Touchable>,
    );

    fireEvent.click(component.container);

    expect(fn).toHaveBeenCalled();
  });

  it('onPressIn', () => {
    const fn = jest.fn();
    const component = render(
      <Touchable onPressIn={fn}>
        <div />
      </Touchable>,
    );

    fireEvent.click(component.container);

    expect(fn).toHaveBeenCalled();
  });

  it('onPressOut', () => {
    const fn = jest.fn();
    const component = render(
      <Touchable onPressOut={fn}>
        <div />
      </Touchable>,
    );

    fireEvent.click(component.container);

    expect(fn).toHaveBeenCalled();
  });
});
