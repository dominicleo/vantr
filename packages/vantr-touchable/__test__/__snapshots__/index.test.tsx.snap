// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Touchable snapshot 1`] = `
<Touchable>
  <div />
</Touchable>
`;
